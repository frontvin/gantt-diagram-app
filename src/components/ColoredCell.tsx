import React from 'react';

export const ColoredCell: React.FC<{ backgroundColor: string }> = ({
  children,
  backgroundColor
}) => {
  return <td style={{ backgroundColor }}>{children}</td>;
};
