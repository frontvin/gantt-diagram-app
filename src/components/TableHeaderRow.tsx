import React from "react";
import { Table } from "semantic-ui-react";

interface ITableHeader {
  tasksHeaderName: string;
  timelineHeaderName: string;
}

export const monthNames = [
  "January",
  "February",
  "March",
  "April",
  "May",
  "June",
  "July",
  "August",
  "September",
  "October",
  "November",
  "December"
];

export const TableHeaderRow: React.FC<ITableHeader> = ({
  tasksHeaderName,
  timelineHeaderName
}: ITableHeader) => {
  return (
    <Table.Header>
      <Table.Row>
        <Table.HeaderCell rowSpan="2" textAlign="center">
          {tasksHeaderName}
        </Table.HeaderCell>
        <Table.HeaderCell colSpan="12" textAlign="center">
          {timelineHeaderName}
        </Table.HeaderCell>
      </Table.Row>
      <Table.Row>
        {monthNames.map((month: string, index: number) => {
          return (
            <Table.HeaderCell key={index} monthname={month} textAlign="center">
              {month}
            </Table.HeaderCell>
          );
        })}
      </Table.Row>
    </Table.Header>
  );
};
