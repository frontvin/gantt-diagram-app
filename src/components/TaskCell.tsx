import React from "react";
import { Table } from "semantic-ui-react";
import { ColoredCell } from "./ColoredCell";

export interface ITaskCell {
  taskName: string;
  monthNumber: number;
  active: boolean;
  cellColor: string;
  getCurrentCell(monthNumber: number, taskName: string): void;
}

export const TaskCell: React.FC<ITaskCell> = ({
     taskName,
     monthNumber,
     active,
     cellColor,
     getCurrentCell
  }) => {
  return (
    <Table.Cell
      key={monthNumber}
      as={ColoredCell}
      backgroundColor={active ? cellColor : ""}
      onClick={() => getCurrentCell(monthNumber, taskName)}
    />
  );
};
