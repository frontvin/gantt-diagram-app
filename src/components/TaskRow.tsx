import React from "react";
import { Table } from "semantic-ui-react";
import { ColoredCell } from "./ColoredCell";

interface ITask {
  taskName: string;
  taskStart: number;
  taskDuration: number;
  cellColor: string;
}

const monthNumbers = Array.from({ length: 12 }, (j, i) => i + 1);

const taskActiveInMonth = (
  monthNumber: number,
  taskStart: number,
  taskDuration: number
): boolean => {
  return taskStart <= monthNumber && monthNumber < taskStart + taskDuration;
};

export function getCurrentCell(): void {
  console.log(monthNumbers);
};

getCurrentCell()

export const TaskRow: React.FC<ITask> = ({
  taskName,
  taskStart,
  taskDuration,
  cellColor
}) => {
  return (
    <Table.Body>
      <Table.Row>
        <Table.Cell>{taskName}</Table.Cell>
        {monthNumbers.map(monthNumber => (
          <Table.Cell
            key={monthNumber}
            as={ColoredCell}
            backgroundColor={
              taskActiveInMonth(monthNumber, taskStart, taskDuration)
                ? cellColor
                : ""
            }
            onClick={getCurrentCell}
          />
        ))}
      </Table.Row>
    </Table.Body>
  );
};
