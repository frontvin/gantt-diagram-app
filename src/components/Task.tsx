import React from 'react';
import { Table } from "semantic-ui-react";
import { TaskCell } from './TaskCell'

// *******************************************************
// Task component
// task interface
interface ITask {
  taskID: number
  taskName: string;
  taskStart: number;
  taskDuration: number;
  cellColor: string;
  changeTaskDuration: () => void
}

const monthNumbers = Array.from({ length: 12 }, (j, i) => i + 1);

const taskActiveInMonth = (
  monthNumber: number,
  taskStart: number,
  taskDuration: number
): boolean => {
  return taskStart <= monthNumber && monthNumber < taskStart + taskDuration;
};

const getCurrentCell = (
  taskID: number,
  monthNumber: number,
): void => {
  // return monthNumber;

  console.log(monthNumber, taskID);

};

export const Task: React.FC<ITask> = (props) => {
  return (
    <Table.Row>
      <Table.Cell>{props.taskName}</Table.Cell>
      {monthNumbers.map(monthNumber => (
        <TaskCell
          key={`${monthNumber}-${props.taskName}`}
          taskID={props.taskID}
          monthNumber={monthNumber}
          taskName={props.taskName}
          taskStart={props.taskStart}
          cellColor={props.cellColor}
          active={taskActiveInMonth(
            monthNumber,
            props.taskStart,
            props.taskDuration
          )}
          getCurrentCell={getCurrentCell}
        />
      ))}
    </Table.Row>
  )
};