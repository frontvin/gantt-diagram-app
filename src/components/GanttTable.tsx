import React, { Component } from "react";
import { Table } from "semantic-ui-react";
import { TableHeaderRow } from "./TableHeaderRow";
import { TaskRow } from "./TaskRow";

export class GanttTable extends Component<{},{}> {
constructor(props: Readonly<{}>){
  super(props);
  this.state = {
    tasks: [
      {
        taskName: "Task 1",
        taskStart: 5,
        taskDuration: 3,
        cellColor: "red"
      },
      {
        taskName: "Task 2",
        taskStart: 1,
        taskDuration: 5,
        cellColor: "blue"
      },
      {
        taskName: "Task 3",
        taskStart: 9,
        taskDuration: 2,
        cellColor: "green"
      },
      {
        taskName: "Task 4",
        taskStart: 3,
        taskDuration: 1,
        cellColor: "yellow"
      },
      {
        taskName: "Task 5",
        taskStart: 1,
        taskDuration: 4,
        cellColor: "purple"
      },
      {
        taskName: "Task 6",
        taskStart: 2,
        taskDuration: 5,
        cellColor: "orange"
      },
      {
        taskName: "Task 7",
        taskStart: 4,
        taskDuration: 2,
        cellColor: "grey"
      },
      {
        taskName: "Task 8",
        taskStart: 3,
        taskDuration: 5,
        cellColor: "blue"
      },
      {
        taskName: "Task 9",
        taskStart: 5,
        taskDuration: 1,
        cellColor: "cyan"
      },
      {
        taskName: "Task 10",
        taskStart: 4,
        taskDuration: 3,
        cellColor: "red"
      }
    ]
  }
}

  render() {
    const { tasks } = this.state.tasks;

    return (
      <Table table="large" columns="13" celled structured>
        <TableHeaderRow
          tasksHeaderName={"Tasks"}
          timelineHeaderName={"Timeline"}
        />
        {tasks.map((task: any , index : any) => {
          return (
            <TaskRow
              key={index}
              taskName={task.taskName}
              taskStart={task.taskStart}
              taskDuration={task.taskDuration}
              cellColor={task.cellColor}
            />
          );
        })}
      </Table>
    );
  }
}
