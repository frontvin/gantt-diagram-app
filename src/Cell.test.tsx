import React from "react";
import renderer from "react-test-renderer";
import { TaskCell } from "./components/TaskCell";

it("renders correctly", () => {
  const CELL = renderer
    .create(
      <TaskCell
        taskID={1}
        monthNumber={3}
        active={true}
        cellColor={"red"}
        taskStart={3}
        getCurrentCell={(taskID, monthNumber) => `taskID ${taskID} + monthNumber ${monthNumber}` }
      />
    )
    .toJSON();
  expect(CELL).toMatchSnapshot();
});
