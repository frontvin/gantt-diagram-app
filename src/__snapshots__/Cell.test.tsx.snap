// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<td
  className=""
  onClick={[Function]}
  style={
    Object {
      "backgroundColor": "red",
    }
  }
/>
`;
